### ISSUE TITLE
Coercion a `Str` to a `Str` takes longer than coercing an `Int` to a `Str`

### ISSUE THREAD

**Issue by lizmat** [rakudo/rakudo#3477]
$ raku -e '"42".Str for ^100000000; say now - INIT now'
    0.96049932

    $ raku -e '42.Str for ^100000000; say now - INIT now'
    0.6685104

This feels... odd.  Not sure whether this is a side effect of something else, but it definitely warrants some closer inspection, as @jnthn suggested on IRC.

**Comment by MasterDuke17** [comment]
Fwiw, I get more intuitive results with non-literals.
```
[dan@alexandria p6]$ perl6 -e 'my $a; my $b = 99.rand.Int.Str; my $s = now; $a = $b.Str for ^50_000_000; say now - $s; say $a'
0.9346032
22
[dan@alexandria p6]$ perl6 -e 'my $a; my $b = 99.rand.Int; my $s = now; $a = $b.Str for ^50_000_000; say now - $s; say $a'
1.6922222
11
```

**Comment by jnthn** [comment]
I took a look. The optimizer is mostly too smart for the original benchmark. In the `42.Str` case, it can tell that the result is unused. It's a pure operation. It deletes the entire coercion logic. So it's not measuring coercion at all, just how few instructions are left behind after all inlining and so on.

The discrepancy is because in the case of the `Str` method on `Int`, it emits one spesh plugin call for return handling; the static optimizer figures out the decont one isn't needed, so it just spits out the type check one. In the `Str.Str` case it can't tell that `self` is a decontainerized symbol, so we end up with a slightly longer code sequence after all inlining - mostly due to the specializer being cautious enough to preserve enough to uninline, and not yet realizing that can never possibly happen. Probably both of these are worth a bit of tuning.

**Comment by jnthn** [comment]
> Probably both of these are worth a bit of tuning.

Currently testing a patch to the static optimizer, which makes the two pretty much the same time.

**Comment by MasterDuke17** [comment]
That makes sense, I get numbers similar in ratio to my non-literal example when I use a literal, but store the result.

**Comment by jnthn** [comment]
> That makes sense, I get numbers similar in ratio to my non-literal example when I use a literal, but store the result.

Yes, we're not smart enough to ruin that benchmark...yet. :-) Ultimately, the original devolved to "how fast is returning `self`", and the answer was "not as fast as we could be", and it's not an unusual thing to do,so I figure it's totally worth saving some cycles there.

**Comment by jnthn** [comment]
Closed with 49d1dcd30c.
